packages/babel-types/src/definitions/core.js

```diff
@@ -329,7 +329,16 @@ export const functionDeclarationCommon = {
 };
 
 defineType("FunctionDeclaration", {
-  builder: ["id", "params", "body", "generator", "async"],
+  builder: [
+    "id",
+    "params",
+    "body",
+    "generator",
+    "async",
+    "declare",
```

Comment:
I see your point, some functions' number of arguments really gets out of hand. Another idea could be to include all important (required) fields as arguments and then have another argument where you can have an object that takes the remaining optional fields:

```js
t.identifier("foo", { typeAnnotation: t.anyTypeAnnotation() });
```

This would give us more of an immutable looking interface. I understand that for example in transformers you want to optimise for speed, on the other hand in my current project I use babel to generate complex type definitions and its complexity reduces significantly with a declarative immutable style.

I can maybe reduce this PR to the fixes (e.g. `name` argument of _TypeParameter_) and open a separate issue for the proposal mentioned above where we can discuss. What do you think?

Reply:
Separate issue sounds good for the proposal!